### DISCUSSION TITLE
Functions for users to move to other instance

### DISCUSSION THREAD

**Discussion by jaschaurbach** [bookwyrm-social/bookwyrm#2368] [Ideas]
I suggest a "move to other instance" protocol in cases a user wants to move and take all information with them.

I use the verbs **MUST** and **SHOULD**. **MUST** is something that needs to be done, **SHOULD** is nice to have but not not necessary for functionality. **USER** is the user, **OLDINSTANCE** is the the instance they are moving away from, **NEWINSTANCE** is the instance they are moving to. **FOLLOWINSTANCE** is the instance a follower of **USER** is on.

For the downloads, these information need at least to be contained in the corresponding files:
`Books`: The persistent identifiers (ISBN13, Openlibrary-ID), reading data (started reading, finished reading, stopped reading), rating, quotes, reviews, shelf.
`Followings`: Handles of all Followings
`Lists` Name of list, which books are on it

*  USER decides to move to another instance. USER **must** download the following information: `Books`, `Followings`, `Lists`, `Blockings` from OLDINSTANCE
* USER **must** create an account on NEWINSTANCE.
* USER **must** enter new account in a field in userprofile on OLDINSTANCE, from this moment on the account **MUST** be set read only and can not be interacted with any more.
* NEWINSTANCE **must** check with OLDINSTANCE if the the new account is set on the profile on OLDINSTANCE. If not, NEWINSTACE **must** do nothing and **should** inform USER that the move can not be executed. After OLDINSTANCE confirmed that the new account is set, the profile on OLDINSTANCE **must** be set to read-only. No more interaction on OLDINSTANCE **must** be possible. OLDINSTANCE **must** show the URL to the new account on NEWINSTANCE.

If the the check validated the move to NEWINSTANCE:
* Followers of USER **must** get informed of the new account name by NEWINSTANCE and **must** start following USER automatically. 
* USER **should** upload all information from OLDINSTANCE on NEWINSTANCE. USER **must** upload at least `Followings`.
* After upload of `Following` USER **must** follow these accounts.
* After upload of `Books` the books are put on the shelfs where they were before on NEWINSTANCE and reading data get imported. Missing shelfs **must** be created on NEWINSTANCE. Missing books **must** get fetched, either imported from OLDINSTANCE or other data providers. USER **should** be able to chose from where the data gets fetched. NEWINSTANCE informs OLDINSTANCE about the import with the new URL. OLDINSTANCE **should** redirect to URL on NEWINSTANCE if clicked from profile view.
* After upload of 'Lists' the lists get created on NEWINSTANCE. NEWINSTANCE informs OLDINSTANCE of new URL, OLDINSTANCE **must** redirect to URL on NEWINSTANCE. Lists USER just follows or contributed get an update from NEWINSTANCE that it checked the move and the account now resides on NEWINSTANCE. 
* After upload of `Blockings` NEWINSTANCE follows the blockings.

I'm sure I forgot something which needs to imported or processed. Please feel free to make other suggestions, additions or even reasons why this should not exist.

**Comment by mouse-reeve** [comment]
Thank you for such thorough thoughts on this! There is a github issue as well: #1012

**Reply by jaschaurbach** [reply]
Oh damned, I had not seen the issue. So maybe close this?
